lighthouse-core/audits/treemap-data.js

```diff
@@ -0,0 +1,315 @@
+   * from a source map, and non-leaf nodes are directories. Leaf nodes have data
+   * for bytes, coverage, etc., when available, and non-leaf nodes have the
+   * same data as the sum of all descendant leaf nodes.
+   * @param {string} sourceRoot
+   * @param {Record<string, SourceData>} sourcesData
+   * @return {Node}
+   */
+  static prepareTreemapNodes(sourceRoot, sourcesData) {
+    /**
+     * @param {string} name
+     * @return {Node}
+     */
+    function newNode(name) {
+      return {
+        name,
+        resourceBytes: 0,
+      };
+    }
+
+    /**
+     * Given a slash-delimited path, traverse the Node structure and increment
+     * the data provided for each node in the chain. Creates nodes as needed.
+     * Ex: path/to/file.js will find or create "path" on `node`, increment the data fields,
+     *     and continue with "to", and so on.
+     * @param {string} source
+     * @param {SourceData} data
+     * @param {Node} node
+     */
+    function addNode(source, data, node) {
+      // Strip off the shared root.
+      const sourcePathSegments = source.replace(sourceRoot, '').split(/\/+/);
+      sourcePathSegments.forEach((sourcePathSegment, i) => {
+        const isLastSegment = i === sourcePathSegments.length - 1;
+
+        let child = node.children && node.children.find(child => child.name === sourcePathSegment);
+        if (!child) {
+          child = newNode(sourcePathSegment);
+          node.children = node.children || [];
+          node.children.push(child);
+        }
+        node = child;
+
+        // Now that we've found or created the next node in the path, apply the data.
+        node.resourceBytes += data.resourceBytes;
+        if (data.unusedBytes) node.unusedBytes = (node.unusedBytes || 0) + data.unusedBytes;
+        if (data.duplicate !== undefined && isLastSegment) {
+          node.duplicate = data.duplicate;
+        }
+      });
+    }
+
+    const rootNode = newNode(sourceRoot);
+
+    // For every source file, apply the data to all components
+    // of the source path, creating nodes as necessary.
+    for (const [source, data] of Object.entries(sourcesData)) {
+      addNode(source || `<unmapped>`, data, rootNode);
+
+      // Apply the data to the rootNode.
+      rootNode.resourceBytes += data.resourceBytes;
+      if (data.unusedBytes) rootNode.unusedBytes = (rootNode.unusedBytes || 0) + data.unusedBytes;
+    }
+
+    /**
+     * Collapse nodes that have only one child.
+     * @param {Node} node
+     */
+    function collapse(node) {
+      while (node.children && node.children.length === 1) {
+        node.name += '/' + node.children[0].name;
+        node.children = node.children[0].children;
+      }
+
+      if (node.children) {
+        for (const child of node.children) {
+          collapse(child);
+        }
+      }
+    }
+    collapse(rootNode);
+
+    // TODO(cjamcl): Should this structure be flattened for space savings?
```

Comment:
> I'd vote to leave it as is without TODOs unless we run into a problem with the canonical structure

yeah, I'd definitely say do a size check for typical and maybe some pathological pages before landing in `default-config`, but better to default to the straightforward encoding if it's reasonable in size.

Reply:
This is after removing fullpage screenshots from experimental; and removed resource summary.

```
coursehero.com, 105 KB aka 22% of .audits
cnn.com, 48 KB aka 6% of .audits
```

Reply:
100K doesn't seem like a showstopper, even if it is suprisingly large just for some file names.

Maybe leave an issue rather than TODO to explore flattening save difference if we ever need some LHR trimming?